drivers/wifi/infineon/airoc_wifi.c

```diff
@@ -0,0 +1,652 @@
+	if (p_whd_result->SSID.length != 0) {
+		p_zy_result->ssid_length = p_whd_result->SSID.length;
+		sprintf(p_zy_result->ssid, "%s", p_whd_result->SSID.value);
+		p_zy_result->channel = p_whd_result->channel;
+		p_zy_result->security = (p_whd_result->security == 0) ? WIFI_SECURITY_TYPE_NONE
+								      : WIFI_SECURITY_TYPE_PSK;
+		p_zy_result->rssi = (int8_t)p_whd_result->signal_strength;
+		p_zy_result->mac_length = 6;
+		memcpy(p_zy_result->mac, &p_whd_result->BSSID, 6);
+	}
+}
+
+static void scan_callback(whd_scan_result_t **result_ptr, void *user_data, whd_scan_status_t status)
+{
+	struct airoc_wifi_data *data = (struct airoc_wifi_data *)user_data;
+	whd_scan_result_t whd_scan_result;
+	struct wifi_scan_result zephyr_scan_result;
+
+	if (status == WHD_SCAN_COMPLETED_SUCCESSFULLY || status == WHD_SCAN_ABORTED) {
+		data->scan_rslt_cb(data->iface, 0, NULL);
+
+		return;
+	}
+	if ((result_ptr != NULL) || (*result_ptr != NULL)) {
+		memcpy(&whd_scan_result, *result_ptr, sizeof(whd_scan_result_t));
+		parse_scan_result(&whd_scan_result, &zephyr_scan_result);
+		data->scan_rslt_cb(data->iface, 0, &zephyr_scan_result);
+	}
+	memset(*result_ptr, 0, sizeof(whd_scan_result_t));
+}
+
+static int airoc_mgmt_scan(const struct device *dev, struct wifi_scan_params *params,
+			   scan_result_cb_t cb)
+{
+	int ret = 0;
+	struct airoc_wifi_data *data = (struct airoc_wifi_data *)dev->data;
+
+	data->scan_rslt_cb = cb;
+
+	if (k_sem_take(&data->sema_common, K_MSEC(AIROC_WIFI_WAIT_SEMA_MS)) != 0) {
+		return -EAGAIN;
+	}
+
+	ret = whd_wifi_scan(airoc_if, params->scan_type, WHD_BSS_TYPE_ANY, &(data->ssid), NULL,
+			    NULL, NULL, scan_callback, &(data->scan_result), data);
+
+	k_sem_give(&data->sema_common);
+	return ret;
+}
+
+static int airoc_mgmt_connect(const struct device *dev, struct wifi_connect_req_params *params)
+{
+	struct airoc_wifi_data *data = (struct airoc_wifi_data *)dev->data;
+	whd_ssid_t ssid = {0};
+	int ret = 0;
+
+	if (k_sem_take(&data->sema_common, K_MSEC(AIROC_WIFI_WAIT_SEMA_MS)) != 0) {
+		return -EAGAIN;
+	}
+
+	if (data->is_sta_connected) {
+		LOG_ERR("Already connected");
+		ret = -EALREADY;
+		goto error;
+	}
+
+	ssid.length = params->ssid_length;
+	memcpy(ssid.value, params->ssid, params->ssid_length);
+
+	whd_scan_result_t scan_result;
+	whd_scan_result_t usr_result = {0};
+
+	usr_result.SSID.length = ssid.length;
+	memcpy(usr_result.SSID.value, ssid.value, ssid.length);
+
+	if (whd_wifi_scan(airoc_if, WHD_SCAN_TYPE_ACTIVE, WHD_BSS_TYPE_ANY, NULL, NULL, NULL, NULL,
+			  scan_cb_search, &scan_result, &(usr_result)) != WHD_SUCCESS) {
+		LOG_ERR("Failed start scan");
+		ret = -EAGAIN;
+		goto error;
+	}
+
+	if (k_sem_take(&airoc_wifi_data.sema_scan, K_MSEC(12 * 1000)) != 0) {
+		whd_wifi_stop_scan(airoc_if);
+		ret = -EAGAIN;
+		goto error;
+	}
+
+	if (usr_result.security == 0) {
+		ret = -EAGAIN;
+		goto error;
+	}
+
+	/* Connect to the network */
+	if (whd_wifi_join(airoc_if, &usr_result.SSID, usr_result.security, params->psk,
```

Comment:
Is this a blocking call till connection succeeds? Don't see any timeout, so, just a single attempt?

Reply:
Correct, It is blocking call (https://github.com/zephyrproject-rtos/hal_infineon/blob/815e84a5150f95627201f192779a0180d5052de7/wifi-host-driver/WiFi_Host_Driver/src/whd_wifi_api.c#L1769), the timeout for join defined in WHD driver as DEFAULT_JOIN_ATTEMPT_TIMEOUT (https://github.com/zephyrproject-rtos/hal_infineon/blob/815e84a5150f95627201f192779a0180d5052de7/wifi-host-driver/WiFi_Host_Driver/src/whd_wifi_api.c#L66C9-L66C42)